components/servo/webview_delegate.rs

```diff
@@ -143,6 +144,90 @@ impl Drop for AuthenticationRequest {
     }
 }
 
+/// Information related to the loading of a web resource. These are created for all HTTP requests.
+/// The client may choose to intercept the load of web resources and send an alternate response
+/// by calling [`WebResourceLoad::intercept`].
+pub struct WebResourceLoad {
+    pub request: WebResourceRequest,
+    pub(crate) response_sender: IpcSender<WebResourceResponseMsg>,
+    pub(crate) intercepted: bool,
+}
+
+impl WebResourceLoad {
+    /// The [`WebResourceRequest`] associated with this [`WebResourceLoad`].
+    pub fn request(&self) -> &WebResourceRequest {
+        &self.request
+    }
+    /// Intercept this [`WebResourceLoad`] and control the response via the returned
+    /// [`InterceptedWebResourceLoad`].
+    pub fn intercept(mut self, response: WebResourceResponse) -> InterceptedWebResourceLoad {
+        let _ = self
+            .response_sender
+            .send(WebResourceResponseMsg::Start(response));
+        self.intercepted = true;
+        InterceptedWebResourceLoad {
+            request: self.request.clone(),
+            response_sender: self.response_sender.clone(),
+            finished: false,
+        }
+    }
+}
+
+impl Drop for WebResourceLoad {
+    fn drop(&mut self) {
+        if !self.intercepted {
+            let _ = self
+                .response_sender
+                .send(WebResourceResponseMsg::DoNotIntercept);
+        }
+    }
+}
+
+/// An intercepted web resource load. This struct allows the client to send an alternatie response
+/// after calling [`WebResourceLoad::intercept`]. In order to send chunks of body data, the client
+/// must call [`InterceptedWebResourceLoad::send_body_data`]. When the interception is complete, the client
+/// should call [`InterceptedWebResourceLoad::finish`]. If neither `finish()` or `cancel()` are called,
+/// this interception will automatically be finished when dropped.
```

Comment:
Not sure if sending `finish` when no explicit action is taken is a safe default e.g memory management bugs on the embedder side can mean the load is finished earlier than expected. `Cancel`ed by default more safe to me, at least assuming it will at the very least show up as an explicit network error.

Reply:
> The only way memory management issues could cause trouble is with unsafe code.

That, and embedders who will integrate with Servo using the C API, which is something we plan to support. I don't see why we need to optimize the number of lines for code here, but I'm fine to leave it as finish-by-default for now.

Reply:
Hrm. I suspect that a C API will expose the Servo types via some kind of reference counting interface (ala GLib). If the reference isn't released, no response will ever be sent. If the reference count is reduced to many times, I expect the C code to segfault because it will be trying to access memory that is already released (but on the embedding side).